{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Chloroplast Curation**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The purpose of this notebook is to describe the identification, assessment and curation of abnormal cds found after automated annotation (using [GeSeq](https://chlorobox.mpimp-golm.mpg.de/geseq.html)) of the _Plukenetia volubilis_ chloroplast sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These analysis were carried out using python and specifically [Biopython](https://biopython.org/), starting from the genbank file `chloroplast_annotated_raw.gb` generated by GeSeq, and ended up in a modified genbank file `chloroplast_annotated_curated.gb`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, some necesary modules are loaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sivico26/miniconda3/envs/Sci/lib/python3.7/site-packages/Bio/Align/substitution_matrices/__init__.py:21: BiopythonExperimentalWarning: Bio.Align.substitution_matrices is an experimental module which may still undergo significant changes. In particular, the location of this module may change, and the Array class defined in this module may be moved to other existing or new modules in Biopython.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqFeature import FeatureLocation, CompoundLocation\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the \"raw\" genbank file is loaded too:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sacha = SeqIO.read(\"chloroplast_annotated_raw.gb\",\"gb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Identification of abnormal cds was executed based on the following criteria:\n",
    "\n",
    "- The sequence must start with a start codon.\n",
    "- The sequence has to have one, and just one, stop codon. Additionally it must be located at the end of the sequence.\n",
    "- The sequence has to be exactly divisible in codons (_i.e._ there should not be remaining nucleotides outside codons)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following functions check the aforementioned conditions and extracts the cds that do not fulfill them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_start(cds, gen_cod):\n",
    "    \"\"\"Return True if the protein starts by a start codon (according to the genetic code)\n",
    "    returns False otherwise\"\"\"\n",
    "    starts = CodonTable.unambiguous_dna_by_id[gen_cod].start_codons\n",
    "    return True if cds.seq[:3] in starts else False\n",
    "\n",
    "def check_exact_cod(cds):\n",
    "    \"Calculate the number of nucleotides outside codons for a protein\"\n",
    "    return len(cds.seq) % 3\n",
    "\n",
    "def check_stops(cds, gen_cod):\n",
    "    \"\"\"Returns the number of stop codons of a protein. \n",
    "    If there is exactly one, instead return True if it is \n",
    "    at the end of que sequence and false otherwise.\"\"\"\n",
    "    \n",
    "    prot = cds.seq.translate(table = gen_cod)\n",
    "    stops = prot.count(\"*\")\n",
    "    if stops != 1:\n",
    "        return stops\n",
    "    else:\n",
    "        return True if prot[-1] == \"*\" else False\n",
    "\n",
    "def make_table(bad_cds):\n",
    "    print(\"index\\tcds\\tstart\\tcod_excess\\tstop\")\n",
    "    for cds, reasons in bad_cds.items():\n",
    "        aux_list = [str(reasons[\"index\"]),cds]\n",
    "        aux_list.append(str(reasons[\"start\"]))\n",
    "        aux_list.append(str(reasons[\"cod\"]))\n",
    "        aux_list.append(str(reasons[\"end\"]))\n",
    "        print(\"\\t\".join(aux_list))\n",
    "    \n",
    "def check_cds(req, gen_cod = 11):\n",
    "    anomalus = {}\n",
    "    for i,ft in enumerate(req.features):\n",
    "        if ft.type == \"CDS\":\n",
    "            reasons = {}\n",
    "            cds = ft.extract(req)\n",
    "            reasons[\"start\"] = check_start(cds, gen_cod)\n",
    "            reasons[\"cod\"] = check_exact_cod(cds)\n",
    "            reasons[\"end\"] = check_stops(cds, gen_cod)\n",
    "            \n",
    "            if reasons[\"start\"] != True or reasons[\"cod\"] != 0 or reasons[\"end\"] != True:\n",
    "                reasons[\"index\"] = i\n",
    "                anomalus[ft.qualifiers[\"gene\"][0]] = reasons\n",
    "    make_table(anomalus)\n",
    "    return anomalus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index\tcds\tstart\tcod_excess\tstop\n",
      "62\trpoC1\tTrue\t1\t0\n",
      "75\tatpF\tTrue\t2\t13\n",
      "99\tndhK\tTrue\t2\t0\n",
      "168\tpetB\tFalse\t1\t13\n",
      "170\tpetD\tFalse\t2\t18\n",
      "178\tinfA\tFalse\t0\t2\n",
      "184\trpl16\tFalse\t1\t10\n",
      "188\trpl22\tTrue\t1\t5\n",
      "243\tycf1\tTrue\t1\tFalse\n",
      "254\tndhI\tTrue\t0\t0\n",
      "256\tndhG\tTrue\t0\t2\n",
      "262\tndhD\tFalse\t0\tTrue\n",
      "264\tccsA\tTrue\t2\t0\n",
      "270\tndhF\tTrue\t1\t0\n",
      "272\tycf1-fragment\tTrue\t2\t0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sivico26/miniconda3/envs/Sci/lib/python3.7/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "anomalus = check_cds(sacha)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus, 15 cds were retrieved, from which at least 14 had to be reviewed (ycf1-fragment is naturally fragmented by the IRs). Separated by kind of anomaly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "anom_start_ft = set()\n",
    "anom_cod_ft = set()\n",
    "anom_end_ft = set()\n",
    "for cds, carac in anomalus.items():\n",
    "    if not carac[\"start\"]:\n",
    "        anom_start_ft.add(cds)\n",
    "    if carac[\"cod\"] != 0:\n",
    "        anom_cod_ft.add(cds)\n",
    "    if carac[\"end\"] is not True:\n",
    "        anom_end_ft.add(cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "11\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "print(len(anom_start_ft))\n",
    "print(len(anom_cod_ft))\n",
    "print(len(anom_end_ft))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In summary (excluding ycf1-fragment):\n",
    "\n",
    "- Proteins without start codon: 5\n",
    "- Proteins with nucleotides outside its codons: 10\n",
    "- Proteins with abnormal Stops: 13"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "anom_ft = [sacha.features[key['index']] for key in anomalus.values()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The starting codon anomalies were addressed first"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Starting codon**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'infA', 'ndhD', 'petB', 'petD', 'rpl16'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anom_start_ft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blatX_hit petB_Rcommunis_GeSeq-SRS_v6, position 3 - 648, psl score 98.3, coverage 99.69%, match 97.99%\n",
      "blatX_hit petD_Hbrasiliensis_GeSeq-SRS_v6, position 8 - 504, psl score 97.4, coverage 98.61%, match 96.03%\n",
      "blatX_hit infA_Vvinifera_GeSeq-SRS_v6, position 57 - 122, psl score 80.4, coverage 28.21%, match 22.65%\n",
      "blatX_hit rpl16_Hbrasiliensis_GeSeq-SRS_v6, position 6 - 408, psl score 96.8, coverage 98.77%, match 95.59%\n",
      "blatX_hit ndhD_Rcommunis_GeSeq-SRS_v6, position 1 - 1503, psl score 96.8, coverage 100.00%, match 96.74%\n"
     ]
    }
   ],
   "source": [
    "for ft in anom_ft:\n",
    "    name = ft.qualifiers[\"gene\"][0]\n",
    "    if name in anom_start_ft:\n",
    "        idx = int(anomalus[name][\"index\"])\n",
    "        start = str(ft.extract(sacha).seq[:3])\n",
    "        if start in CodonTable.unambiguous_dna_by_id[11].start_codons:\n",
    "            print(f'{start} from {name} is a start codon')\n",
    "        else:\n",
    "            print(sacha.features[idx - 1].qualifiers[\"note\"][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Something that all these genes have in common is that they were annotated __by reference__ using __blatX__. Also, all except ndhD, did not start in the first nucleotide of their reference, which might explain why they were missing their start codons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also the match with infA has less than a third of the protein coverage, so it is unlikely that this protein is functional."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Finding the missing start codons__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the cds, even though partial, were real, the starting codons would not be far from annotated. Thus, the starting codons were searched upstream and downstream"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def offset_loc(location, offset):\n",
    "    \"\"\"Given a locaton in a feature location, return an offseted location by 'offset' bp (integer number)\n",
    "    Here, a positive value of the offset would return a a location moved FORWARD, while a negative\n",
    "    value would return a location moved BACKWARD\"\"\"\n",
    "    if offset == 0:\n",
    "        raise ValueError(f\"Expected an offset different from 0\")\n",
    "    elif isinstance(location, FeatureLocation):\n",
    "        return location + location.strand * offset\n",
    "    elif isinstance(location, CompoundLocation):\n",
    "        return sum(part + part.strand * offset for part in location.parts)\n",
    "    else:\n",
    "        raise TypeError(f\"Expected either FeatureLocation or CompoundLocation object ({type(location)} given)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_start(feature, mother, starts, forward = False, start_offset = 1):\n",
    "    \"\"\"Given a feature(cds) and its mother sequence search for a codon, present \n",
    "    in starts. Start_offset modifies the location where of the feature before the search.\n",
    "    If forward is True the search is made downstream the beginning of the sequence, \n",
    "    otherwise it is made upstream. \n",
    "    \n",
    "    Returns how many positions, either upstream or downstream the starting codon was found.\n",
    "    \"\"\"\n",
    "    start = feature.extract(mother).seq[:3]\n",
    "    name = feature.qualifiers[\"gene\"][0]\n",
    "    if start in starts:\n",
    "        print(f\"The start {start} of {name} is a start codon\")\n",
    "    elif forward:\n",
    "        i = start_offset if start_offset > 0 else start_offset * -1\n",
    "        missing_start = True\n",
    "        while missing_start:\n",
    "            start = offset_loc(feature.location, i).extract(mother).seq[:3]\n",
    "            if start in starts:\n",
    "                print(f\"A start codon {start} was found {i} bp downstream the sequence for {name}\")\n",
    "                missing_start = False\n",
    "                return i \n",
    "            i += 1\n",
    "    else:\n",
    "        i = start_offset if start_offset < 0 else start_offset * -1\n",
    "        missing_start = True\n",
    "        while missing_start:\n",
    "            start = offset_loc(feature.location, i).extract(mother).seq[:3]\n",
    "            if start in starts:\n",
    "                print(f\"A start codon {start} was found {i} bp upstream the sequence for {name}\")\n",
    "                missing_start = False\n",
    "                return i\n",
    "            i -= 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Upstream:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A start codon ATC was found -4 bp upstream the sequence for petB\n",
      "A start codon ATA was found -1 bp upstream the sequence for petD\n",
      "A start codon ATG was found -7 bp upstream the sequence for infA\n",
      "A start codon ATA was found -1 bp upstream the sequence for rpl16\n",
      "A start codon TTG was found -11 bp upstream the sequence for ndhD\n"
     ]
    }
   ],
   "source": [
    "for ft in anom_ft:\n",
    "    name = ft.qualifiers[\"gene\"][0]\n",
    "    if name in anom_start_ft:\n",
    "        find_start(ft, sacha, CodonTable.unambiguous_dna_by_id[11].start_codons)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downstream:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A start codon ATA was found 2 bp downstream the sequence for petB\n",
      "A start codon CTG was found 12 bp downstream the sequence for petD\n",
      "A start codon ATG was found 1 bp downstream the sequence for infA\n",
      "A start codon ATT was found 18 bp downstream the sequence for rpl16\n",
      "A start codon ATT was found 4 bp downstream the sequence for ndhD\n"
     ]
    }
   ],
   "source": [
    "for ft in anom_ft:\n",
    "    name = ft.qualifiers[\"gene\"][0]\n",
    "    if name in anom_start_ft:\n",
    "        find_start(ft, sacha, CodonTable.unambiguous_dna_by_id[11].start_codons, forward = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As suspected, starting codons were not far either upstream or downstream. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ndhD was quite particular, as in three other cases (_[Jatroha curcas](https://www.ncbi.nlm.nih.gov/nucleotide/NC_012224.1), [Nicotiana tabacum](https://www.ncbi.nlm.nih.gov/nucleotide/NC_001879.2) and [Arabidopsis thaliana](https://www.ncbi.nlm.nih.gov/nucleotide/NC_000932.1)_) it starts by ACG, even though ACG is not a start codon for the bacterial genetic code. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __Finding a reasonable start__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that possible starts for protein were located, the next step is to assess if taking those gives a reasonable protein. \n",
    "\n",
    "One way to do so is to compare with a trustful reference, and align the modified proteins with their respective homologues. Usually, the closer the reference the better. However, as _Ricinus communis_ chloroplast had presented issues before, for these analysis the model organism _[Arabidopsis thaliana](https://www.ncbi.nlm.nih.gov/nucleotide/NC_000932.1)_ was preferred."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"svillanu@eafit.edu.co\"\n",
    "\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=\"NC_000932\", rettype=\"gb\", retmode=\"text\")\n",
    "arabidopsis = SeqIO.read(handle, \"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_protein(feature, mother, ref, starts, gen_cod, reach = 50):\n",
    "    \"\"\"Take a feature with its mother sequence and look for alternatives \n",
    "    based in the starts given, within a reach upstream and downstream. \n",
    "    \n",
    "    Returns a dictionary where the keys are the offset relative to the start\n",
    "    of the protein and the values are the alignment of the modified protein \n",
    "    with the homologue of the reference\"\"\"\n",
    "    \n",
    "    name = feature.qualifiers[\"gene\"][0]\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    aligner.open_gap_score = -5 \n",
    "    aligner.extend_gap_score = -0.5\n",
    "    aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "    ## To Do: are these good alignment defaults?\n",
    "    \n",
    "    \n",
    "    found = False\n",
    "    for ft in ref.features:\n",
    "        if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(ref).seq.translate(table = gen_cod)\n",
    "            found = True\n",
    "    \n",
    "    if not found:\n",
    "        raise NameError(f\"Protein {name} could not be found in the provided reference\") \n",
    "    \n",
    "    alns = {}\n",
    "    max_score = -1000 ## To Do: look for a better default\n",
    "    idx_max = 0 \n",
    "    i = 1\n",
    "    while abs(i) < reach:\n",
    "        i = find_start(feature, mother, starts, forward=True, start_offset=i)\n",
    "        query = offset_loc(feature.location, i).extract(mother).seq.translate(table = gen_cod)\n",
    "        alns[i] = aligner.align(target, query)\n",
    "        if alns[i].score > max_score:\n",
    "            max_score = alns[i].score\n",
    "            idx_max = i\n",
    "        i += 1\n",
    "        \n",
    "    i = -1\n",
    "    while abs(i) < reach:\n",
    "        i = find_start(feature, mother, starts, forward=False, start_offset=i)\n",
    "        query = offset_loc(feature.location, i).extract(mother).seq.translate(table = gen_cod)\n",
    "        alns[i] = aligner.align(target, query)\n",
    "        if alns[i].score > max_score:\n",
    "            max_score = alns[i].score\n",
    "            idx_max = i\n",
    "        i -= 1\n",
    "    print(f\"A total of {len(alns.keys())} options were found in the range of {reach}\")\n",
    "    print(f\"The best option found in the explored range was found with an offset of {idx_max} with an score of {max_score}\")\n",
    "    return alns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "petB was taken as an example of the above function. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A start codon ATA was found 2 bp downstream the sequence for petB\n",
      "A start codon ATG was found 11 bp downstream the sequence for petB\n",
      "A start codon ATT was found 14 bp downstream the sequence for petB\n",
      "A start codon TTG was found 15 bp downstream the sequence for petB\n",
      "A start codon ATT was found 37 bp downstream the sequence for petB\n",
      "A start codon ATT was found 46 bp downstream the sequence for petB\n",
      "A start codon TTG was found 47 bp downstream the sequence for petB\n",
      "A start codon ATG was found 53 bp downstream the sequence for petB\n",
      "A start codon ATC was found -4 bp upstream the sequence for petB\n",
      "A start codon CTG was found -15 bp upstream the sequence for petB\n",
      "A start codon ATA was found -41 bp upstream the sequence for petB\n",
      "A start codon ATA was found -43 bp upstream the sequence for petB\n",
      "A start codon ATT was found -49 bp upstream the sequence for petB\n",
      "A total of 13 options were found in the range of 50\n",
      "The best option found in the explored range was found with an offset of -41 with an score of 1020.5\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"petB\"][\"index\"]\n",
    "alns = find_protein(sacha.features[idx], sacha, arabidopsis, CodonTable.unambiguous_dna_by_id[11].start_codons, 11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking at the score of these alignments:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(2, 57.5),\n",
       " (11, 61.5),\n",
       " (14, 57.5),\n",
       " (15, 81.0),\n",
       " (37, 1014.5),\n",
       " (46, 998.5),\n",
       " (47, 40.5),\n",
       " (53, 40.0),\n",
       " (-4, 64.5),\n",
       " (-15, 78.5),\n",
       " (-41, 1020.5),\n",
       " (-43, 62.5),\n",
       " (-49, 67.5)]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(i,aln.score) for i,aln in alns.items()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first number of the tuple is the offset where the start codon was found (a positive value implies it was found downstream, while a negative value means it was found upstream).The second number of the tuple is the score of the alignment.\n",
    "\n",
    "Based on the scores, 37 and 46 downstream and 41 upstream were reasonalbe candidates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Finding stop codons**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the start candidates are set, the stop codons have to be located if missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_stop(feature, mother, gen_cod, offset = 0):\n",
    "    \"\"\"Given a feature and its mother sequence search for \n",
    "    the next stop codon if there is none. Also the nucleotides that\n",
    "    do not belong to any codon are trimmed from the end of the sequence.\n",
    "    \n",
    "    If succesful, it returns a modified location that fulfills cds requiriments\"\"\"\n",
    "    # To do: modify to use with CompoundLocations with offset \n",
    "    if offset != 0:\n",
    "        loc = feature.location + offset * feature.location.strand ## This only applies to FeatureLocations ::: To Do: generalize to CompoundLocation\n",
    "    else:\n",
    "        loc = feature.location\n",
    "    prot = loc.extract(mother).seq.translate(table=gen_cod)\n",
    "    stops = prot.count(\"*\")\n",
    "    if stops != 0:\n",
    "        print(f\"This feature already has {stops} stop codons\")\n",
    "        if stops == 1 and prot[-1] == \"*\":\n",
    "            print(f\"This protein has already a unique stop codon in the end of the sequence\")\n",
    "        elif stops == 1:\n",
    "            print(f\"And it is located at position {prot.find('*')} of {len(prot)}\")\n",
    "        else:\n",
    "            print(f\"And the first of them is located at position {prot.find('*')} of {len(prot)}\")\n",
    "    else:\n",
    "        # Going downstream\n",
    "        name = feature.qualifiers[\"gene\"][0]\n",
    "        not_found = True\n",
    "        i = 0\n",
    "        excess = len(loc) % 3 \n",
    "        if isinstance(loc, FeatureLocation):\n",
    "            if excess != 0:\n",
    "                if loc.strand == 1:\n",
    "                    loc = FeatureLocation(loc.start, loc.end - excess*loc.strand, loc.strand)\n",
    "                elif loc.strand == -1:\n",
    "                    loc = FeatureLocation(loc.start - excess*loc.strand, loc.end, loc.strand)\n",
    "            while not_found:\n",
    "                i += 1\n",
    "                if loc.strand == 1:\n",
    "                    loc = FeatureLocation(loc.start, loc.end + 3*loc.strand, loc.strand)\n",
    "                elif loc.strand == -1:\n",
    "                    loc = FeatureLocation(loc.start + 3*loc.strand, loc.end, loc.strand)\n",
    "                    \n",
    "                if loc.extract(mother).seq.translate(table=gen_cod).endswith(\"*\"):\n",
    "                    print(f\"A stop for {name} was found {i} codons downstream ({i*3} bp) and the protein length would be {int(len(loc)/3)} aa\")\n",
    "                    not_found = False\n",
    "        elif isinstance(loc, CompoundLocation): # To Do: modify it to work with (-) locations (and heterogeneous?)\n",
    "            end_exon = loc.parts[-1] \n",
    "            if excess != 0:\n",
    "                if end_exon.strand == 1:\n",
    "                    end_exon = FeatureLocation(end_exon.start, end_exon.end - excess*end_exon.strand, end_exon.strand)\n",
    "                elif end_exon.strand == -1:\n",
    "                    end_exon = FeatureLocation(end_exon.start - excess*end_exon.strand, end_exon.end, end_exon.strand)\n",
    "            while not_found:\n",
    "                i += 1\n",
    "                end_exon = FeatureLocation(end_exon.start, end_exon.end + 3*loc.strand, end_exon.strand)\n",
    "                loc = sum(part for part in loc.parts[:-1]) + end_exon\n",
    "                if loc.extract(mother).seq.translate(table=gen_cod)[-1] == \"*\":\n",
    "                    print(f\"A stop for {name} was found {i} codons downstream ({i*3} bp) and the protein length would be {int(len(loc)/3)} aa\")\n",
    "                    not_found = False\n",
    "        else:\n",
    "            raise TypeError(f\"Expected either FeatureLocation or CompoundLocation object ({type(location)} given)\")\n",
    "        return loc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Continuing the petB example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A stop for petB was found 14 codons downstream (42 bp) and the protein length would be 229 aa\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"petB\"][\"index\"]\n",
    "new_petb = find_stop(sacha.features[idx], sacha, 11, -41)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('MRFSEGEFLWFTYLNKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITL...GPL', ExtendedIUPACProtein())"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_petb.extract(sacha).seq.translate(table=11, cds=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above protein fulfilled the cds characteristics, but did it have biological sense? An alignment to the reference shed some light."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M-------------SKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLGVLTASFGVTGYSLPWDQIGYWAVKIVTGVPDAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFLMIRKQGISGPL*\n",
      "|-------------.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||-\n",
      "MRFSEGEFLWFTYLNKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLSVLTASFGVTGYSLPWDQIGYWAVKIVTGVPEAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFPMIRKQGISGPL-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "aligner = Align.PairwiseAligner()\n",
    "aligner.open_gap_score = -5 \n",
    "aligner.extend_gap_score = -0.5\n",
    "aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "for ft in arabidopsis.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == \"petB\":\n",
    "        target = ft.extract(arabidopsis).seq.translate(table = 11)\n",
    "\n",
    "query = new_petb.extract(sacha).seq.translate(table=11, cds=True)\n",
    "print(aligner.align(target, query)[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It seems as the best cds achievable for this gene. But does that insertion goes there? further assessment should be done to be more sure. If it is compared with _[Ricinus communis](https://www.ncbi.nlm.nih.gov/nucleotide/NC_016736.1)_ as reference:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "handle = Entrez.efetch(db=\"nucleotide\", id=\"NC_016736\", rettype=\"gb\", retmode=\"text\")\n",
    "ricinus = SeqIO.read(handle, \"gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MRFSEGEALWFTYLNKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLAVLTASFGVTGYSLPWDQIGYWAVKIVTGVPEAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFPMIRKQGISGPL\n",
      "|||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "MRFSEGEFLWFTYLNKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLSVLTASFGVTGYSLPWDQIGYWAVKIVTGVPEAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFPMIRKQGISGPL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "aligner = Align.PairwiseAligner()\n",
    "aligner.open_gap_score = -5 \n",
    "aligner.extend_gap_score = -0.5\n",
    "aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "for ft in ricinus.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == \"petB\":\n",
    "        target = ft.extract(ricinus).seq.translate(table = 11, cds=True)\n",
    "\n",
    "query = new_petb.extract(sacha).seq.translate(table=11, cds=True)\n",
    "print(aligner.align(target, query)[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As the insertion is shared with _Ricinus communis_ it might be considered natural."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Fixing the remaining starting proteins__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again the cds with starts anomalies are:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'infA', 'ndhD', 'petB', 'petD', 'rpl16'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anom_start_ft"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As aforementioned, infA can be discarded because its coverage is too low to be functional. Additionally, besides the difference in the start codon, ndhD has no cds problems."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TNDFPWLTIIVVFPISAGSLMLFLPHRGNKVNKWYTICICILELLLTTYAFCYNFKMDDPLIQLSEDYKWIDFFDFYWRMGIDGLSIGTILLTGFITTLATLAAFPVTRDSRFFHFLMLAMYSGQIGSFSSRDLLLFFIMWELELIPVYLLLSMWGGKKRLYSATKFILYTAGSSIFLLIGVLGISLYGSNEPTLNLELLANKSYPVTLEILFYIGFLIAFAVKSPIIPLHTWLPDTHGEAHYSTCMLLAGILLKMGAYGLVRINMELLPHAHSMFSPWLLVVGTIQIIYAASTSPGQRNLKKRIAYSSVSHMGFIIIGISSITDPGLNGAILQIISHGFIGAALFFLAGTSYDRIRLVYLDEMGGMAISIPKIFTMFTILSMASLALPGMSGFIAEFIVFFGIITSQKYFLISKIFIIFVMAIGMILTPIYLLSMLRQMFYGYKLINIKNFSFFDSGPRELFLSISILLPIIGIGIYPDFVLSLASDKVESILSNYFYG*\n",
      "||.||||||.||.|.|||||..|.||.||||..|||..|||||.||.||.|.|.|..|||||||.||||||.|||||||.||||.|...|||||||||||||||.|.|||||.||||||||||||||.|||.|||||||||||||||||||||.|||||||||||||||||||.|.|||.|.|||.|||||||....|..||.||||.|||.|||||||||||||||||||.||||||||||||||||||||||||||||||||||||||||||.|||||...|.|||||||||||||||||.|||||||||||||||||.||.|.|.|||||||||||||||||||||||||||||.||||||||||...|||||.|.|||.|||||||||||.||...||||||.|||.|.|||.|.|.||.||||||||||||||||||||||....|..||||||||||.|||||||.|||||||||..||..|.||..||||||.|\n",
      "TNSFPWLTIFVVLPVSAGSLIFFFPHKGNKVIRWYTLFICILEFLLLTYVFFYYFQLDDPLIQLTEDYKWIQFFDFYWRLGIDGFSLIPILLTGFITTLATLAARPITRDSRLFHFLMLAMYSGQIGLFSSQDLLLFFIMWELELIPVYLLLSVWGGKKRLYSATKFILYTAGGSVFLLMGALGIALYGSNEPRFHFETSANQSYPVALEIFFYIGFLIAFAVKSPIIPLHIWLPDTHGEAHYSTCMLLAGILLKMGAYGLVRINMELLPHAHSTFSPWLIIIGVIQIIYAASTSPGQRNLKQRIAYSSVSHMGFIIIGIYSINDMGFNGAILQIISHGFIGAALFFLAGTSYDRIRRVYLDEMGGMAALMPKIFTTFSILSLASLALPGMSGFFAELAIFFGIITGQKYLLMSKIVITFGMAVGMILTPIYLLSMLRQMFYGYKLFTAPNSYFFDSGPRELFVSISILLPVIGIGIYPDFLFSLSVDRVEALLSNYFYK*\n",
      "\n"
     ]
    }
   ],
   "source": [
    "aligner = Align.PairwiseAligner()\n",
    "aligner.open_gap_score = -5 \n",
    "aligner.extend_gap_score = -0.5\n",
    "aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "\n",
    "for ft in arabidopsis.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == \"ndhD\":\n",
    "        target = ft.extract(arabidopsis).seq.translate(table = 11)\n",
    "\n",
    "query = sacha.features[anomalus[\"ndhD\"][\"index\"]].extract(sacha).seq.translate(table=11)\n",
    "print(aligner.align(target, query)[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hence both were excluded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "anom_start_ft_rem = anom_start_ft.copy()\n",
    "for prot in [\"ndhD\",\"infA\"]:\n",
    "    anom_start_ft_rem.discard(prot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'petB', 'petD', 'rpl16'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anom_start_ft_rem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the above process with petB is replicated for the other start anomalous cds the result is the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginnig with petB\n",
      "A start codon ATA was found 2 bp downstream the sequence for petB\n",
      "A start codon ATG was found 11 bp downstream the sequence for petB\n",
      "A start codon ATT was found 14 bp downstream the sequence for petB\n",
      "A start codon TTG was found 15 bp downstream the sequence for petB\n",
      "A start codon ATT was found 37 bp downstream the sequence for petB\n",
      "A start codon ATT was found 46 bp downstream the sequence for petB\n",
      "A start codon TTG was found 47 bp downstream the sequence for petB\n",
      "A start codon ATG was found 53 bp downstream the sequence for petB\n",
      "A start codon ATC was found -4 bp upstream the sequence for petB\n",
      "A start codon CTG was found -15 bp upstream the sequence for petB\n",
      "A start codon ATA was found -41 bp upstream the sequence for petB\n",
      "A start codon ATA was found -43 bp upstream the sequence for petB\n",
      "A start codon ATT was found -49 bp upstream the sequence for petB\n",
      "A total of 13 options were found in the range of 50\n",
      "The best option found in the explored range was found with an offset of -41 with an score of 1020.5\n",
      "A total of 3 candidates were found for petB\n",
      "This feature already has 3 stop codons\n",
      "And the first of them is located at position 202 of 215\n",
      "This feature already has 3 stop codons\n",
      "And the first of them is located at position 199 of 215\n",
      "A stop for petB was found 14 codons downstream (42 bp) and the protein length would be 229 aa\n",
      "Beginnig with petD\n",
      "A start codon CTG was found 12 bp downstream the sequence for petD\n",
      "A start codon TTG was found 17 bp downstream the sequence for petD\n",
      "A start codon ATG was found 21 bp downstream the sequence for petD\n",
      "A start codon ATC was found 24 bp downstream the sequence for petD\n",
      "A start codon CTG was found 27 bp downstream the sequence for petD\n",
      "A start codon ATT was found 31 bp downstream the sequence for petD\n",
      "A start codon ATT was found 43 bp downstream the sequence for petD\n",
      "A start codon TTG was found 44 bp downstream the sequence for petD\n",
      "A start codon ATG was found 56 bp downstream the sequence for petD\n",
      "A start codon ATA was found -1 bp upstream the sequence for petD\n",
      "A start codon ATC was found -7 bp upstream the sequence for petD\n",
      "A start codon ATT was found -15 bp upstream the sequence for petD\n",
      "A start codon ATA was found -21 bp upstream the sequence for petD\n",
      "A start codon ATT was found -25 bp upstream the sequence for petD\n",
      "A start codon ATG was found -29 bp upstream the sequence for petD\n",
      "A start codon ATG was found -49 bp upstream the sequence for petD\n",
      "A total of 16 options were found in the range of 50\n",
      "The best option found in the explored range was found with an offset of -7 with an score of 815.5\n",
      "A total of 7 candidates were found for petD\n",
      "This feature already has 2 stop codons\n",
      "And the first of them is located at position 152 of 165\n",
      "This feature already has 3 stop codons\n",
      "And the first of them is located at position 143 of 165\n",
      "This feature already has 4 stop codons\n",
      "And the first of them is located at position 139 of 165\n",
      "This feature already has 1 stop codons\n",
      "And it is located at position 158 of 165\n",
      "This feature already has 1 stop codons\n",
      "And it is located at position 160 of 165\n",
      "A stop for petD was found 2 codons downstream (6 bp) and the protein length would be 167 aa\n",
      "A stop for petD was found 10 codons downstream (30 bp) and the protein length would be 175 aa\n",
      "Beginnig with rpl16\n",
      "A start codon ATT was found 18 bp downstream the sequence for rpl16\n",
      "A start codon ATA was found 32 bp downstream the sequence for rpl16\n",
      "A start codon ATG was found 43 bp downstream the sequence for rpl16\n",
      "A start codon ATA was found 52 bp downstream the sequence for rpl16\n",
      "A start codon ATA was found -1 bp upstream the sequence for rpl16\n",
      "A start codon ATC was found -8 bp upstream the sequence for rpl16\n",
      "A start codon TTG was found -21 bp upstream the sequence for rpl16\n",
      "A start codon ATT was found -22 bp upstream the sequence for rpl16\n",
      "A start codon ATG was found -27 bp upstream the sequence for rpl16\n",
      "A start codon CTG was found -38 bp upstream the sequence for rpl16\n",
      "A start codon ATG was found -48 bp upstream the sequence for rpl16\n",
      "A start codon ATG was found -64 bp upstream the sequence for rpl16\n",
      "A total of 12 options were found in the range of 50\n",
      "The best option found in the explored range was found with an offset of -8 with an score of 623.0\n",
      "A total of 4 candidates were found for rpl16\n",
      "This feature already has 1 stop codons\n",
      "And it is located at position 119 of 134\n",
      "This feature already has 1 stop codons\n",
      "And it is located at position 116 of 134\n",
      "A stop for rpl16 was found 3 codons downstream (9 bp) and the protein length would be 137 aa\n",
      "This feature already has 2 stop codons\n",
      "And the first of them is located at position 2 of 134\n"
     ]
    }
   ],
   "source": [
    "alt_cand = {}\n",
    "alt_loc = {}\n",
    "starts = CodonTable.unambiguous_dna_by_id[11].start_codons\n",
    "for ft in anom_ft:\n",
    "    name = ft.qualifiers[\"gene\"][0]\n",
    "    if name in anom_start_ft_rem:\n",
    "        print(f\"Beginnig with {name}\")\n",
    "        alns = find_protein(ft, sacha, arabidopsis, starts, 11, reach = 50)\n",
    "        dict_scores = {i:aln.score for i,aln in alns.items()}\n",
    "        avg_score = sum(dict_scores.values()) / len(dict_scores)\n",
    "        aln_candidates = {i:aln[0] for i, aln in alns.items() if aln.score > avg_score} ## The filtering criterion could be improved\n",
    "        print(f\"A total of {len(aln_candidates)} candidates were found for {name}\")\n",
    "        alt_cand[name] = aln_candidates\n",
    "        \n",
    "        alt_loc[name] = {}\n",
    "        for i, aln in aln_candidates.items():\n",
    "            alt_loc[name][i] = find_stop(ft, sacha, 11, offset = i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'petB': {37: None,\n",
       "  46: None,\n",
       "  -41: FeatureLocation(ExactPosition(81013), ExactPosition(81700), strand=1)},\n",
       " 'petD': {17: None,\n",
       "  44: None,\n",
       "  56: None,\n",
       "  -1: None,\n",
       "  -7: None,\n",
       "  -25: FeatureLocation(ExactPosition(82687), ExactPosition(83188), strand=1),\n",
       "  -49: FeatureLocation(ExactPosition(82663), ExactPosition(83188), strand=1)},\n",
       " 'rpl16': {43: None,\n",
       "  52: None,\n",
       "  -8: FeatureLocation(ExactPosition(86871), ExactPosition(87282), strand=-1),\n",
       "  -38: None}}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_loc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a result, 4 alternatives were obtained for 3 proteins (petD has 2). \n",
    "A final alignment with the reference to check the results:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __petB__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"petB\"\n",
    "off = -41\n",
    "ref = arabidopsis\n",
    "for ft in ref.features:\n",
    "    if ft.type == \"CDS\":\n",
    "        if ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(arabidopsis).seq.translate(table=11, cds=True)\n",
    "\n",
    "query = alt_loc[name][off].extract(sacha).seq.translate(table=11, cds=True)\n",
    "alns = aligner.align(target, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M-------------SKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLGVLTASFGVTGYSLPWDQIGYWAVKIVTGVPDAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFLMIRKQGISGPL\n",
      "|-------------.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||\n",
      "MRFSEGEFLWFTYLNKVYDWFEERLEIQAIADDITSKYVPPHVNIFYCLGGITLTCFLVQVATGFAMTFYYRPTVTEAFASVQYIMTEANFGWLIRSVHRWSASMMVLMMILHVFRVYLTGGFKKPRELTWVTGVVLSVLTASFGVTGYSLPWDQIGYWAVKIVTGVPEAIPVIGSPLVELLRGSASVGQSTLTRFYSLHTFVLPLLTAVFMLMHFPMIRKQGISGPL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __rpl16__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"rpl16\"\n",
    "off = -8\n",
    "ref = arabidopsis\n",
    "for ft in ref.features:\n",
    "    if ft.type == \"CDS\":\n",
    "        if ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(arabidopsis).seq.translate(table=11, cds=True)\n",
    "\n",
    "query = alt_loc[name][off].extract(sacha).seq.translate(table=11, cds=True)\n",
    "alns = aligner.align(target, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M-LSPKRTRFRKQHRGRLKGISSRGNRICFGRYALQTLEPAWITSRQIEAGRRAMTRNVRRGGKIWVRIFPDKPVTVRPAETRMGSGKGSPEYWVAVVKPGKILYEMGGVPENIARKAISIAASKMPIKTQFIISE\n",
      "|-..|||||||||||||.|||..|||||||||||||.|||||||||||||||||||||.|||||||||||||||||.||.||||||||||||||||.||||.||||||||.|||||.|||||||||||.||||||.\n",
      "MNYNPKRTRFRKQHRGRMKGIAFRGNRICFGRYALQALEPAWITSRQIEAGRRAMTRNARRGGKIWVRIFPDKPVTLRPTETRMGSGKGSPEYWVAIVKPGRILYEMGGVSENIARRAISIAASKMPIRTQFIISG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __petD__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the case of petD that has 2 options, these were compared first between them to choose one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"petD\"\n",
    "off = -25\n",
    "ref = arabidopsis\n",
    "for ft in ref.features:\n",
    "    if ft.type == \"CDS\":\n",
    "        if ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(arabidopsis).seq.translate(table=11, cds=True)\n",
    "\n",
    "query = alt_loc[name][off].extract(sacha).seq.translate(table=11, cds=True)\n",
    "alns = aligner.align(target, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M------GVTKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLIGTAAALWLGIGATLPIDKSLTLGLF\n",
      "|------..|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||||||\n",
      "MYKNSPIQITKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLIGTAVALWLGIGATLPIDKSLTLGLF\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"petD\"\n",
    "off = -49\n",
    "ref = arabidopsis\n",
    "for ft in ref.features:\n",
    "    if ft.type == \"CDS\":\n",
    "        if ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(arabidopsis).seq.translate(table=11, cds=True)\n",
    "\n",
    "query = alt_loc[name][off].extract(sacha).seq.translate(table=11, cds=True)\n",
    "alns = aligner.align(target, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M-----G---------VTKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLIGTAAALWLGIGATLPIDKSLTLGLF\n",
      "|-----|---------.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||||||\n",
      "MSGSFGGWIYKNSPIQITKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLIGTAVALWLGIGATLPIDKSLTLGLF\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As this case was similar to the petB case, the same strategy was used: compare to a closer reference (_Ricinus communis_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "name = \"petD\"\n",
    "off = -49\n",
    "ref = ricinus\n",
    "for ft in ref.features:\n",
    "    if ft.type == \"CDS\":\n",
    "        if ft.qualifiers[\"gene\"][0] == name:\n",
    "            target = ft.extract(ref).seq.translate(table=11)\n",
    "\n",
    "query = alt_loc[name][off].extract(sacha).seq.translate(table=11)\n",
    "alns = aligner.align(target, query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSGSFGGWIYKNSPIPITKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLVGTAVALWLGIGATLPIDKSLTLGLF*\n",
      "|||||||||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||\n",
      "MSGSFGGWIYKNSPIQITKKPDLNDPVLRAKLAKGMGHNYYGEPAWPNDLLYIFPVVILGTIACNVGLAVLEPSMIGEPADPFATPLEILPEWYFFPVFQILRTVPNKLLGVLLMVSVPAGLLTVPFLENVNKFQNPFRRPVATTVFLIGTAVALWLGIGATLPIDKSLTLGLF*\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As the -49 offset matched almost perfectly,  it is the chosen candidate. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Remaining CDS__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After the 5 cds that have abnormal starting sites were adressed, there were still 9 cds to review with excess or stop codon problems. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cds that remained which ones do remain?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "rem_ft = {ft.qualifiers[\"gene\"][0] for ft in anom_ft} - anom_start_ft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rpoC1 {'start': True, 'cod': 1, 'end': 0, 'index': 62}\n",
      "atpF {'start': True, 'cod': 2, 'end': 13, 'index': 75}\n",
      "ndhK {'start': True, 'cod': 2, 'end': 0, 'index': 99}\n",
      "rpl22 {'start': True, 'cod': 1, 'end': 5, 'index': 188}\n",
      "ycf1 {'start': True, 'cod': 1, 'end': False, 'index': 243}\n",
      "ndhI {'start': True, 'cod': 0, 'end': 0, 'index': 254}\n",
      "ndhG {'start': True, 'cod': 0, 'end': 2, 'index': 256}\n",
      "ccsA {'start': True, 'cod': 2, 'end': 0, 'index': 264}\n",
      "ndhF {'start': True, 'cod': 1, 'end': 0, 'index': 270}\n",
      "ycf1-fragment {'start': True, 'cod': 2, 'end': 0, 'index': 272}\n"
     ]
    }
   ],
   "source": [
    "for key, value in anomalus.items():\n",
    "    if key in rem_ft:\n",
    "        print(key,value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarly as the previous cds, modified proteins were proposed after searching for missing stop codons,  but no offset was needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginnig with rpoC1\n",
      "A stop for rpoC1 was found 3 codons downstream (9 bp) and the protein length would be 683 aa\n",
      "Beginnig with atpF\n",
      "This feature already has 13 stop codons\n",
      "And the first of them is located at position 48 of 184\n",
      "Beginnig with ndhK\n",
      "A stop for ndhK was found 1 codons downstream (3 bp) and the protein length would be 226 aa\n",
      "Beginnig with rpl22\n",
      "This feature already has 5 stop codons\n",
      "And the first of them is located at position 145 of 185\n",
      "Beginnig with ycf1\n",
      "This feature already has 1 stop codons\n",
      "And it is located at position 1897 of 1904\n",
      "Beginnig with ndhI\n",
      "A stop for ndhI was found 7 codons downstream (21 bp) and the protein length would be 172 aa\n",
      "Beginnig with ndhG\n",
      "This feature already has 2 stop codons\n",
      "And the first of them is located at position 176 of 178\n",
      "Beginnig with ccsA\n",
      "A stop for ccsA was found 2 codons downstream (6 bp) and the protein length would be 322 aa\n",
      "Beginnig with ndhF\n",
      "A stop for ndhF was found 1 codons downstream (3 bp) and the protein length would be 752 aa\n",
      "Beginnig with ycf1-fragment\n",
      "A stop for ycf1-fragment was found 21 codons downstream (63 bp) and the protein length would be 659 aa\n"
     ]
    }
   ],
   "source": [
    "for ft in anom_ft:\n",
    "    name = ft.qualifiers[\"gene\"][0]\n",
    "    if name in rem_ft:\n",
    "        print(f\"Beginnig with {name}\")\n",
    "        alt_loc[name] = {}\n",
    "        alt_loc[name][0] = find_stop(ft, sacha, 11, offset = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although stop codons were found for some of the remaining proteins (rpoC1, ndhK, ndhI, ccsA, ndhF), others need further assessment such as atpF, rpl22, ycf1, ndhG as they have multiple stop codons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The alignments for those cds that were fixed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment for rpoC1 query 682 target 680\n",
      "MIDRYKHQQLRIGLVSPQQISAWATKIIPNGEIVGEVTKPYTFHYKTNKPEKDGLFCERIFGPIKSGICACGNYRVIGDEKEDPKFCEQCGVEFVDSRIRRYQMGYIKLTCPVTHVWYLKRLPSYIANLLDKPLKELEGLVYCDFSFARPITKKPTFLRLRGSFEYEIQSWKYSIPLFFTTQGFDIFRNREISTGAGAIREQLADLDLRIIIENSLVEWKQLGEEGPTGNEWEDRKIVRRKDFLVRRMELAKHFIRTNIEPEWMVLCLLPVLPPELRPIIQIEGGKLMSSDINELYRRVIYRNNTLTDLLTTSRSTPGELVMCQEKLVQEAVDTLLDNGIRGQPMRDGHNKVYKSFSDVIEGKEGRFRETLLGKRVDYSGRSVIVVGPSLSLHRCGLPREIAIELFQTFVIRGLIRQHLASNIGVAKSQIREKKPIVWEILQEVMQGHPVLLNRAPTLHRLGIQSFQPILVEGRTICLHPLVCKGFNADFDGDQMAVHVPLSLEAQAEARLLMFSHMNLLSPAIGDPISVPTQDMLIGLYVLTSGTRRGICANRYNPCNRKNYQNERIYETNYKYTKEPFFCNSYDAIGAYRQKKINLDSPLWLRWQLDQRVIASREVPIEVHYESFGNYHEIYAHYLIVRSVKKENFCIYIRTTVGHISFYREIEEAIQGFSQACS--YDT\n",
      "||||||||||||||||||||||||.||.|||||||||||||||||||||||||||||||||||||||||||||||||..|||.|||||||||||||||.||||||||||.|||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||.||||.||||||||||||||.|||||||||||..|.||||.|||.|||||||||||..|||||||||.||||||||||||||||||||||||||||||||||.|||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||||||||.||||||||||||||||||||.||||.||||||||||||||||||||||||||||||.|||||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|.|||||||||||..|.|||||...|..||||.||.||||||||||||.|||||.||||||||||.|||||.|.||||||.|.|||||.|.||||..|||..|||||||||||..|||||||||||.||.|--...\n",
      "MIDRYKHQQLRIGLVSPQQISAWANKILPNGEIVGEVTKPYTFHYKTNKPEKDGLFCERIFGPIKSGICACGNYRVIRNEKEKPKFCEQCGVEFVDSRVRRYQMGYIKLACPVTHVWYLKRLPSYIANLLDKPLKELEGLVYCDFSFARPIAKKPTFLRLRGSFEYEIQSWKYSIPLFFTAQGFDTFRNREISTGAGAIRKQLADLDLRIIIDYSAVEWKELGEKGPTGNEWEDRKVGRRKDFLVRRVELAKHFIRTNIEPEWMVLCLLPVLPPELRPIIQIDGGKLMSSDINELYRRVIYRNNTLIDLLTTSRSTPGELVMCQEKLVQEAVDTLLDNGIRGQPMRDGHNKVYKSFSDVIEGKEGRFRETMLGKRVDYSGRSVIVVGPSLSLHRCGLPREIAIELFQIFVIRGLIRQHLASNIGVAKSKIREKEPIVWEILQEVMQGHPVLLNRAPTLHRLGIQAFQPILVEGRAICLHPLVCKGFNADFDGDQMAVHVPLSLEAQAEARLLMFSHMNLLSPAIGDPISVPTQDMLIGLYVLTSGNRQGICANRYNPCNHRNFQNERIDNNNDQYTKESFFSNSYDAIGAYRQKRINLDSSLWLRWQLDQRAIASREAPVEVHYESLGTYHEIYEHLLIVRNIKKEILCIYIRTTVGHIYLYREIEEAIQGFCQAGSNGIES\n",
      "\n",
      "Alignment for ndhK query 225 target 225\n",
      "MNSIKFPILDRTTKNSVISTTLNDLSNWSRLSSLWPLLYGTSCCFIEFASLIGSRFDFDRYGLVPRSSPRQADLILTAGTVTMKMAPSLVRLYEQMPEPKYVIAMGACTITGGMFSTDSYSTVRGVDKLIPVDVYLPGCPPKPEAVIDAITKLRKKIAREIYKDRIRPQQGNRCFTTNHKFFVVRSPHIGNYDQELLY-PPS-STSEISTETFFKY-KSPVSSHELVN\n",
      "||.|.||.|||||..||||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.|||||..||||.||||.|.|||||||||||......|.|||||.|||-|||-|||.|..||||||-|.|---.||||\n",
      "MNYIEFPLLDRTTQISVISTTSNDLSNWSRLSSLWPLLYGTSCCFIEFASLIGSRFDFDRYGLVPRSSPRQADLILTAGTVTMKMAPSLVRLYEQMPEPKYVIAMGACTITGGMFSTDSYSTVRGVDKLIPVDVYLPGCPPKPEAVIDAIIKLRKKLSREIYEDRIRSQPGNRCFTTNHKFNIEPATHTGNYDQGLLYQPPSTSTSKIPSETFFKYNKRP---PELVN\n",
      "\n",
      "Alignment for ndhI query 171 target 172\n",
      "MLPMITGFMNYGQQTLRAARYIGQGFMITLSHTNRLPVTIQYPYEKLITSERFRGRIHFEFDKCIACEVCVRVCPIDLPVVDWKLETNIRKKRLLNYSIDFGICIFCGNCVEYCPTNCLSMTEEYEFSTYDRHELNYNQIALGRLPMSVIDDYTIRTIWNSPQTKNGVN-PLI\n",
      "|.|..||||||||||.||||||||.|.|||||.||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||.||||||||.||...|.-..-||-\n",
      "MFPVVTGFMNYGQQTIRAARYIGQSFIITLSHANRLPVTIQYPYEKLITSERFRGRIHFEFDKCIACEVCVRVCPIDLPVVDWKLETAIRKKRLLNYSIDFGICIFCGNCVEYCPTNCLSMTEEYELSTYDRHELNYNQIALGRLPMSVVDDYTIRTILNSTKKKK-IDKPL-\n",
      "\n",
      "Alignment for ccsA query 321 target 328\n",
      "MIFSILEHILTHISFSVVSIVLTIYFLTLLVNLDEIIGFFDSSDKGIIITFFGITGLLLTRWIYSGHFPLSNLYESLIFLSWAFSIIHMVSYFNKKQQNKLNTITAPSVIFIQGFATSGLLNKMPQSAILVPALQSQWLMMHVSMMILGYGALLCGSLLSIALLVITFRKVGPTFWKKNI--KKNFLLNELFSFDVLYYINERNSILLQQNINFSFSRNYYRYQLIQQLDFWSFRIISLGFIFLTVGILSGAVWANETWGSYWNWDPKETWAFITWTIFAIYLHIKTNRNVRGINSAIVALIGFILIWICYFGVNLLGIGLHSYGSF--TSN\n",
      "|||..||||||||||||.|.|.||...||.|--||.....|||.||.|.|||.|||||.|||...||.||||||||||||||||||||||.||-||..|.|..|||||..|.|||||||||....||..||||||||||||||||||.||.|||||||||.|.|.|||||----|..--|--|.|.|||..|||....|..||...|--.||.|..||||||.|||||||.||.||||.||||||.|||||||||||.||||||||||||||||||||||||.|..||....|.||||||..||..|||||||||||||||||||||--|||\n",
      "MIFATLEHILTHISFSVMSLVITIHLITLVV--DEFVELYDSSEKGMITTFFCITGLLVTRWGFGGHLPLSNLYESLIFLSWAFSIIHMVPYF-KKHKNFLSAITAPSTFFTQGFATSGLLTNTHQSKTLVPALQSQWLMMHVSMMIFGYAALLCGSLLSVAFLIITFRK----FIR--IVDKSNNLLNASFSFSKIQYMTERKNVL--RNISFLSSRNYYRFQLIQQLDDWSYRIISIGFIFLTIGILSGAVWANEAWGSYWNWDPKETWAFITWTIFAIYFHTRTNKKFEGFNSAIVASMGFLIIWICYFGVNLLGIGLHSYGSFTLTSN\n",
      "\n",
      "Alignment for ndhF query 751 target 746\n",
      "MEHTYQYSWIIPFIPLPVPILLGVGLLLFPTATKNLRRMWTFLSIFLLSIVMIFSIYLSIQQIFLSCIHQNVWSWTINNEFSFEFGYFIDPLTSIMSILITTVGILVLIYSDNYMSHDQGYLRFFAYMGFFNTSMLGLVTSSNLIQVYFFWELVGMCSYLLIGFWFTRPIAANACQKAFVTNRVGDFGLLLGILGLYWITGSFEFQDLFEIFNNLILNNRVNLLFLTLCAFLLFVGPIAKSAQFPLHVWLPDAMEGPTPISALIHAATMVAAGIFLVARLLPLFIVIPSIMYIISLIGIITVLLGATLALAQKDIKRGLAYSTMSQLGYMMLALGMGSYRSALFHLITHAYSKALLFLGSGSIIHSMEAIVGYSPDKSQNMILMGGLTKHVPITKTAFLIGTLSLCGIPPLACFWSKDEILNDSLLFSPIFAIIACSTAGLTAFYMFRIYLLTFEGHLNTYFLNYSGKKSGSFYSLSLWGKEEEKKLNKNFGLVPLLTMNNTKRASFFCNKTYKISNNVRNQI--FITVENF---GLNT-RTFYYPHESDNTILFPMLILVLFTLFIGAIGIPFNQEGIDFDILSKFFTPSINLLHKNSQNFVDWYEFLRNATFSVSIAFFGIFIAYCLYKPFYSSLLNLTLLNSFQKWNS------KRIHWEKLINFVYNWSYNRGYIDSFFKTSLIESIRRLAKQTTFFDKRIIDGITNGVGITSFFVGEVTKYIGGSRISSYLFLYLSYVLIFLMILFFFYFEKF\n",
      "|||.|||||.|||...|||||.|.||||||.|||.|||||.|.|.|||.|||||||.|||.|...|...|..|.|||||.||||||...||||||.|.|||||||.||.||||||||||||||||.||.|||||||||.|||||||.|.||||||||||||||||||||||.|||||||||||||||||||||||.||||||.||.|||.|.|||..||.|..||.||||||.|.|.|.|||||||||||||||||||||||||||||||||||||||||.|||.|||.||..|.|||||||.||||||||||||||.||||||||||||||||||||||.|||||||||||||||||||||||||||.||||.|||||||.||||||||.|||||.|..||||||||||.|||||||||||||.|.|||||||||.|||||||||||.|||.||||||..|.|..|||..||||.||||||..|..|.|||....||.||..|||||...||.|..|||...--|||.-|.---..||-|.|.||||||||.|||||.||||.|||||.|||||.||...|||||..|||.||||.|..|..|||||..|..|||||||||||||...|||.||||.|---||-|--|||------|||.|.|.||..|||||||||||.|.....|..||.||....||||||||||||||...||||||..||.|..||||||...|...||||.|..||.|...\n",
      "MEHIYQYSWVIPFVTFPVPILIGAGLLLFPAATKRLRRMWAFPSLFLLTIVMIFSINLSIHQLNSSFSYQYIWLWTINNDFSFEFGHLLDPLTSILSVLITTVGIMVLFYSDNYMSHDQGYLRFFSYMSFFNTSMLGLITSSNLIQIYIFWELVGMCSYLLIGFWFTRPIASNACQKAFVTNRVGDFGLLLGILGIYWITGSLEFRDLFQILNNLVYNNQVHFLFVTLCAFLVFSGAISKSAQFPLHVWLPDAMEGPTPISALIHAATMVAAGIFLVARLFPLFVVIPYIMNLIALIGIITVFLGATLALAQKDIKRSLAYSTMSQLGYMMLALGMGSYRVALFHLITHAYSKALLFLGSGSIIHSMETIVGYPPDKSQNMVLMGGLTKHIPITKTTFFFGTLSLCGIPPFACFWSKDEILNDSWLYSPIFAIIACFTAGLTAFYMFRVYLLIFEGHLNVHFQNFIGKKNSSFYSISLWGKEGSKMFNQNFGVLTFLTTNNNQRASFFWKNTYQIDDNVRKMMWPFITI-NITRKNCNTKRIFSYPHESDNTMLFPMLVLVLFILFIGALGIPFNHEGMEMDILSKLLTPSLNLLHQNPRNSLDWYEFATNSIFSVSIAFFGIFIASSFYKPAYSSLQN---LN-F--WNSLAKKGPKRIRWDKIINVIYNWSYNRGYIDAFYEIFFIGGIRKLAELIYFFDKRIIDGITNGVSVISFFVGEGIKYVGRGRISSYLLFFLFFTLIFLLIFSFFSFYSL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for name, dict_loc in alt_loc.items():\n",
    "    if 0 in dict_loc.keys() and dict_loc[0]:\n",
    "        if name == \"ycf1-fragment\": \n",
    "            continue\n",
    "        else:\n",
    "            query = dict_loc[0].extract(sacha).seq.translate(table=11, cds=True)\n",
    "\n",
    "            for ft in arabidopsis.features:\n",
    "                if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "                    target = ft.extract(arabidopsis).seq.translate(table=11, cds=True)\n",
    "\n",
    "            alns = aligner.align(target, query)\n",
    "            print(f\"Alignment for {name} query {len(query)} target {len(target)}\")\n",
    "            print(alns[0])\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All the alternatives seem to be appropiate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Focusing again on the ones with multiple stop codons they were adressed individually. As they have multiple stop codons, it was likely that the reading frame was not the appropiate. Hence translation and alignment in different reading frames was tested for these cds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __atpF__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment with offset 0 and score 246.5\n",
      "MKNLTDSFVYLGHWPSAGSFGFNTDILATNPINLSVVFGVLIFFGKGV--LNDLLDNRKQRIL-----NTIR---NSEELREG-AI-----QQL--E-------NAR--------ARLRN---VETEADKFRVNGYSEIEREKL----N-LINSTYKTLKQL------ENYKNETILFEQQRTINQVRERVFQQALQGAIGTLNSCLSNELHLRTINANIGMFGTMKE--ITD*\n",
      "|||.|||||.|||||||.|||||||||||||||||||.||||||||||--......|....||-----|..|---||..-..|-|.-----.|.--|-------|.|--------.||.|---.-|....|..|-----..|.|----|-..|..|..|...------..|-------------------......-----....||----------------|..|.--|---\n",
      "MKNVTDSFVSLGHWPSAESFGFNTDILATNPINLSVVLGVLIFFGKGV*LIY*IIENKRFWILFEIQKNYARGPLNSWK-KPGPAYGKWK*KQISFE*MDTRR*NERN*I*LIQLIRLWNN*KI-TKMKPFILN-----NKERLIKSDNGFSNKPYREL*EF*IVV*PTSY-------------------IYVPSM-----LILACL----------------GR*KK*LI---\n",
      "\n",
      "Alignment with offset -1 and score 65.0\n",
      "MKNLTDSFVYLGHWP-SAGSFGFNTDILATNPINLSVVFGVLI-FFGKGVLNDL-------------LD--------NRKQRIL--------------------NTIRNSEEL------REGAIQQLENARARLRNVET---------EA---DKF----RVNG-YSEI----E--------REKL--NLINSTYKTL---------KQ-----LENY-KN--------ETILFE----------QQR---T---INQ-----VRER-VFQQALQGAIGTLNSCLSNELHL---RTIN--------ANIG-MFGTMKEI---TD*\n",
      "-----------|--|-|...||----.|----.|...--|...-||------|.-------------||--------.|...||--------------------|||.....|------||--.|.|.-----|..||.---------|.---..|----|...-|.||----|--------.|.|--..|.|....|---------.|-----...|-.|--------..|||.----------|..---|---...-----.||.-.|.|--|--------||---.||---|...--------...|-...|.|.|---|.-\n",
      "-----------G--PYSIHLFG----LL----LNREM--GLIVHFF------DIIYKASSG*FKSKQLDVRLGPI*HDRSIEILQHSTFVIYSIHHTR*ISYSWNTIHFQDALVVKW*TRE--TQNLV-----LKSVEVRVLFKA*Y*ECSFNEQFNKRSRIESVYMEIF*CSE**FWLLAVQEFLFRQFIPSIHSVLI*DFNSSMFQQ*HIVPCPKYGRNRCFHDSTTQSILFHLIPISWPHIFQAKEWETFLLLHESNFHFIREKPSFSQ--Q--------CL---CHLIQ*RSVR*EQI**ILITLG*SIRTEKSIR**TI-\n",
      "\n",
      "Alignment with offset -2 and score 56.0\n",
      "MKNLTDSFVYLGHWPSAGS--------FGFN----------------TDILATNPIN-------LSVVFGV-----LIF---------FGKGVLNDLLDNR--KQRILNTIRNSEELREGAIQQLENARARLR---NVETE-ADKFRVNG-Y------------SEIEREKL----NLINS---------------TYKTLKQLENYKNETILFEQ-QRTINQVRERVFQQA--LQG-AIGTLN-----------------SC--LS----------NELHLRTINANIGMF--GTMKEITD*\n",
      "-----....|...|.|..|--------|..|----------------|....|.|||-------|...|..-----.||---------.|-|---...|.|--|.|-------..|-|.|.-..|.....|..---....|-.|....||-.------------|.|----|----..|..---------------||......--.|..--.|..-....|.|.......|--|.|-..|.|.-----------------|.--||----------..|-..|.|..|-..--|....|...\n",
      "-----GTLFYSPLWASIESRNGFDCPFF*YNI*GIFRIIQIEAIGCPTRAYMT*PINRNTPTLHLCHIFHTSH*IDIIFMEYDSLSRCLG-G---EMVDTRDSKSR-------AKE-RGGS-SPLQGIILRMLIQ*AIQ*EISDRIGINGNFLM**VIVLVARCSRI----LV*AVHTIHT*CFDLRFQFFHVSAVTYCSMSKI--WKKQ--VFPRLYHPVNSVPLNPYFMATYLSG*GMGNLSPIT*IQFSFHPGKAIFFSTMPLSFDPIAFR*IGTDL-INTDNSRI-EY*NGKIH*IMNY\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"atpF\"][\"index\"]\n",
    "ft = sacha.features[idx]\n",
    "name = ft.qualifiers[\"gene\"][0]\n",
    "\n",
    "for i in range(3):\n",
    "    query = (ft.location + -i).extract(sacha).seq.translate(table=11)\n",
    "    \n",
    "    for ft in arabidopsis.features:\n",
    "        if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "                target = ft.extract(arabidopsis).seq.translate(table=11)\n",
    "                \n",
    "    alns = aligner.align(target,query)\n",
    "    print(f\"Alignment with offset {-1*i} and score {alns.score}\")\n",
    "    print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although the first alignment had the highest score, it was not considerably higher than the other ones. Furthermore, that same alignment had a high similarity at the start of the sequence which seemed different from the middle and the end of the sequence. Further inspection revealed that this cds was composed by a short exon and other longer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[144, 410]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = anomalus[\"atpF\"][\"index\"]\n",
    "list(map(len,sacha.features[idx].location.parts))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Evaluating the reading frames for the exons individually:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon 1\n",
      "Has 0 stop codons in reading frame 0\n",
      "Has 1 stop codons in reading frame -1\n",
      "Has 7 stop codons in reading frame -2\n",
      "exon 2\n",
      "Has 13 stop codons in reading frame 0\n",
      "Has 0 stop codons in reading frame -1\n",
      "Has 10 stop codons in reading frame -2\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"atpF\"][\"index\"]\n",
    "ex1, ex2 = sacha.features[idx].location.parts\n",
    "\n",
    "print(\"exon 1\")\n",
    "for i in range(0,-3,-1):\n",
    "    count = (ex1 + i).extract(sacha).seq.translate(table=11).count(\"*\")\n",
    "    print(f\"Has {count} stop codons in reading frame {i}\")\n",
    "    \n",
    "print(\"exon 2\")\n",
    "for i in range(0,-3,-1):\n",
    "    count = (ex2 + i).extract(sacha).seq.translate(table=11).count(\"*\")\n",
    "    print(f\"Has {count} stop codons in reading frame {i}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The second exon was not reading the appropiate frame, although it still misses the stop codon. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment with offset -1 in the last exon and score 820.0\n",
      "MKNLTDSFVYLGHWPSAGSFGFNTDILATNPINLSVVFGVLIFFGKGVLNDLLDNRKQRILNTIRNSEELREGAIQQLENARARLRNVETEADKFRVNGYSEIEREKLNLINSTYKTLKQLENYKNETILFEQQRTINQVRERVFQQALQGAIGTLNSCLSNELHLRTINANIGMFGTMKEITD*\n",
      "|||.|||||.|||||||.|||||||||||||||||||.|||||||||||.||||||||.||.|||||||||||||.|||.||||||.||.|||.||.|||||||||||||||||||||.||||||||||.|||||||||||.||||||||||.|.|||||.|||||||||||.|||||.||||.-\n",
      "MKNVTDSFVSLGHWPSAESFGFNTDILATNPINLSVVLGVLIFFGKGVLTDLLDNRKQKILDTIRNSEELREGAIEQLEKARARLRKVEIEADQFRMNGYSEIEREKLNLINSTYKTLEQLENYKNETIHFEQQRTINQVRQRVFQQALQGALGILNSCLTNELHLRTINANLGMFGTIKEITY-\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"atpF\"][\"index\"]\n",
    "name = sacha.features[idx].qualifiers[\"gene\"][0]\n",
    "exon = sacha.features[idx].location.parts[-1] + -1\n",
    "query = (sacha.features[idx].location.parts[0] + exon).extract(sacha).seq.translate(table=11)\n",
    "\n",
    "for ft in arabidopsis.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "        target = ft.extract(arabidopsis).seq.translate(table=11)\n",
    "\n",
    "alns = aligner.align(target,query)\n",
    "print(f\"Alignment with offset -{1} in the last exon and score {alns.score}\")\n",
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It just misses the stop codon:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment score 826.0\n",
      "MKNLTDSFVYLGHWPSAGSFGFNTDILATNPINLSVVFGVLIFFGKGVLNDLLDNRKQRILNTIRNSEELREGAIQQLENARARLRNVETEADKFRVNGYSEIEREKLNLINSTYKTLKQLENYKNETILFEQQRTINQVRERVFQQALQGAIGTLNSCLSNELHLRTINANIGMFGTMKEITD*\n",
      "|||.|||||.|||||||.|||||||||||||||||||.|||||||||||.||||||||.||.|||||||||||||.|||.||||||.||.|||.||.|||||||||||||||||||||.||||||||||.|||||||||||.||||||||||.|.|||||.|||||||||||.|||||.||||.|\n",
      "MKNVTDSFVSLGHWPSAESFGFNTDILATNPINLSVVLGVLIFFGKGVLTDLLDNRKQKILDTIRNSEELREGAIEQLEKARARLRKVEIEADQFRMNGYSEIEREKLNLINSTYKTLEQLENYKNETIHFEQQRTINQVRQRVFQQALQGALGILNSCLTNELHLRTINANLGMFGTIKEITY*\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"atpF\"][\"index\"]\n",
    "exon = sacha.features[idx].location.parts[-1] \n",
    "loc = (sacha.features[idx].location.parts[0] + FeatureLocation(exon.start + -1, exon.end, exon.strand))\n",
    "query = loc.extract(sacha).seq.translate(table=11)\n",
    "\n",
    "for ft in arabidopsis.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "        target = ft.extract(arabidopsis).seq.translate(table=11)\n",
    "        \n",
    "alns = aligner.align(target,query)\n",
    "print(f\"Alignment score {alns.score}\")\n",
    "print(alns[0])\n",
    "\n",
    "alt_loc[\"atpF\"] = {0:loc}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __rpl22__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As rpl22 had several stop codons, its reading frame was checked first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Has 5 stop codons in reading frame 0\n",
      "the first in 145 of 185\n",
      "Has 20 stop codons in reading frame -1\n",
      "the first in 0 of 185\n",
      "Has 9 stop codons in reading frame -2\n",
      "the first in 2 of 185\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"rpl22\"][\"index\"]\n",
    "aux = sacha.features[idx].location\n",
    "\n",
    "for i in range(0,-3,-1):\n",
    "    prot = (aux + i).extract(sacha).seq.translate(table=11)\n",
    "    count = prot.count(\"*\")\n",
    "    print(f\"Has {count} stop codons in reading frame {i}\") \n",
    "    print(f\"the first in {prot.find('*')} of {len(prot)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The reading frame was the correct one. However, If the start codon and the reading frame were already set, the only thing that could be done in this case was to set the end of the cds where the first stop codon is found "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def early_stop(feature, mother, gen_cod):\n",
    "    \"\"\"Given a feature and its mother sequence that has an early\n",
    "    stop codon, it returns a modified location with the corrected\n",
    "    end at the first stop codon\"\"\"\n",
    "    excess = len(feature.location) % 3\n",
    "    orig_len = len(feature.location.extract(mother).seq.translate(table=gen_cod))\n",
    "    stop_len = len(feature.location.extract(mother).seq.translate(table=gen_cod, to_stop = True))\n",
    "    mod = (orig_len - stop_len - 1)*3 + excess\n",
    "    loc = feature.location\n",
    "    if loc.strand == -1:\n",
    "        mod_loc = FeatureLocation(loc.start - mod*loc.strand , loc.end, loc.strand)\n",
    "    elif loc.strand == 1:\n",
    "        mod_loc = FeatureLocation(loc.start, loc.end - mod*loc.strand, loc.strand)\n",
    "\n",
    "    #print(mod_loc.extract(sacha).seq.translate(gen_cod, cds=True))\n",
    "    return mod_loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment and score 607.5 - target (160) - query (145)\n",
      "MINKRKK-----KR---DPYTEVYALGQHICMSTHKARRIIDQIRGRSYEETLMILELMPYRACYPILKLIYSAAANARHNMGFNEANLIISKAEVNEGATVKKLKPQARGRGYLIKRSTCHITIVLKNISLYEEYEEYEEYNICLKNRGWIKKTKSTDATFHDMYSK\n",
      "|||||||-----|.---||||.||||||||.||||||||||.||||||||||||||||||||||||||||||||||||||||||||..|||||||||||||||||||||||||||||||||||||||||||||||...|..|.|-|----------------------\n",
      "MINKRKKGDPYTKEYALDPYTTVYALGQHISMSTHKARRIIEQIRGRSYEETLMILELMPYRACYPILKLIYSAAANARHNMGFNETDLIISKAEVNEGATVKKLKPQARGRGYLIKRSTCHITIVLKNISLYEESDGYSKYSI-L----------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"rpl22\"][\"index\"]\n",
    "name = sacha.features[idx].qualifiers[\"gene\"][0]\n",
    "\n",
    "query = early_stop(sacha.features[idx], sacha, 11).extract(sacha).seq.translate(table=11, cds=True)\n",
    "#query = sacha.features[idx].extract(sacha).seq.translate(table=11)\n",
    "\n",
    "for ft in ricinus.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "        target = ft.extract(ricinus).seq.translate(table=11, cds=True)\n",
    "\n",
    "alns = aligner.align(target,query)\n",
    "print(f\"Alignment and score {alns.score} - target ({len(target)}) - query ({len(query)})\")\n",
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_loc[\"rpl22\"] = {0:early_stop(sacha.features[idx],sacha, 11)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the early stop, further assessment would be required to be sure if the cds is functional."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __ycf1__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment and score 7991.0 - target (1806) - query (1904)\n",
      "MIFKSFILGNLVSLCMKIINLVVVVGLYYGFLTTFSMGPSYLFLLRARVIEEGEEGTEKKVSATTGFITGQLMMFISIYYAPLHLALGRPHTITVLALPYLLFHFFWNNHKHFFDYGATTRNSMRNLSIQFVFLNNLIFQLFNHFILPSSMLVRLVNIYMFRCNNKMLFVTSSFVGWLIGHILFMKWVGLILVWIQQNNSIRSNVLFRSNKYLVSELRNSMARIFSILLFITCVYSLGRIPSPIFTKKLKETSETEEREESEEETDVEIETTSETKGTKQGSTEEDPSSSLFSEEKEDRDKIDETEEIQVNGKEKTKDEFHFHFNETCYKNRPLYETFYLDGNQENSKLEILIEKKKKNLFWFEKPLVTILFDSKRWNRPLRYIKNDQFENAVRNEMSQYFFYTCRSDGKERISFTYPPSLSTFFEMVQKKIFLFTTTTEKLSSDELYNHWNYKNEQKKKNLSNEFINRVQALDKGYLVLNTLEKRTRLCNDKTKKEYLPKIYDPLLSGSYRGRTKFFFSPSTLNINKTSIKNSTEMVWINKIHLILLISNYIEFEPKTDRKSFSTEIAYFLNLINEFAGKSTSSFNFKGLPLFPDYKEEKMDLELLQILKFVFDPVLANSKNKTIKNNSL----GIKEINKQVPHWSYKLIDDLEQQEGENEENMAEDYEIRSRKAKRVVIFTDNQENTATYNNTKDTTDFDQIDEVTLIRYSQQSDFRRDIIKGSTRAQRRKIAIFELFQANAHSPLFLDRIDKSLFFSFDIFELIKTMFINWMCKNAEFTISDYTYT--EKKTKESKKKDEDTREDDKREERTRIEIAEAWDSILFAQIIRGCILVTQSILRKYIILPLLIIAKNCIRILFFQIPEWSEDLKDWSREMHIKCTYNGVQLSEKEFPKNWLTDGIQIKILFPFRLKPWHRSKLKFPHKDPMKKKVQRNDFCFLTVLGMEAELPFGFPRKRLSLFEPIFKKLEKKIIKMKKKGFRIRRTLEERRKLFLNLSKEKANWVIQNSFFRKDIINKINKFEKSKRNRILLSGVKEVYELNATKKEKDSLINKNNRASQKMSTS-TRSMVWTNFSLTEKKMKDLSARRKIILNQIEKISKEKEKKIITSEINISPNKISYNAKTLKLKSSKNISQILKRRNAQLAHKFHFFIKILIERIYIDIFLGIINIPRINAQLFLESTKKIVTKYIYNNEENKKKIDKTNQNTIHFISIINIKKSLNNSNIDVINKNSQILCDISSLSQAYVFYKLSQIKIINLYKLRSILQYHGLFLKNEIKDYFRVQGLFNFEL---KEKNVRNSVMNQWKNWLRSHYQYKYDLSQIRWSRLIPQKWRNRINQYQHPMVQNKKLNKWNLYEKDKLIHYEKKKDFEADSLPNQKDNFKKHYKYNLLSYKSINYENKKDSYIYGSPLQI-NNKQKISYNYNTNKSNFFDMLEGIPINNYLAEDDIIDMAKSPDRKYFDWKILNFCLREKVEIESWIDTGSEHKKNTKTTKYQIIEKMDKKKIFFITIHQDQEVNSSNQK----------------------------------------------------------------------------------------------------------------KDIEKDYMELDMKKHRHKKKNKSHTEVELDFFLKRYLYFQLRWNGSLNQKIINNIKVYCLLLKLTNPREIIISSIQRQEISLNILMIQKDLTLTELMKKRIFTIEPIRLSVKNNGQFLLYQTISILLFHKNKQQINQQYRDKIDSI---SIERQQGIIGNRDKNGYDLLVPENVLSPKRRRELRILISFNLKNKNNINIYINTKICTGNNKKNCGPVFDKSKDFDKEKNKLIKFKFFLWPHFRLEDLACMNRYWFDTNNSSRFGMVRICIYPRLKF---CNGKGAFFLYPIGYLI*\n",
      "||.|||||||||||||||.|.||||||||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||.|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||||||||||||||||||||||||||||||||||||---|||||||||||||||||||||.|||||||||||||||.|||||||||||||||||||||||||.||.||||||||||||||||||||||||||||||||||||||||||||||||||||.||||||.|.|||||||||||||||||||||..|||||.|||||.||||.|.|||--|||||.|.|||||||.|||||||||||||..|||||||||||||||.||||||||||||||||||||||||||||....||||||||--||||||||||||||||||||||||||||||||.|||||||||||.|||||||||||.|||.|||.|||||||||||.|||||||||||||||||||||||..||.----|||||.|.||.||||.|...||||.||||||.|..|||||||||||||||||.||||..||||||.||||||||||||||.|||||||||||.|||||||||||||||||||||||||..|||||||.|.||||||||||||||.||.|||.|||--||||.|.||||||.|||||||||||||||||||||||.|.||..|||||||||||||||.|||.||.||||||||||||||||||..|||.||||||||||||||||||||||||||||||||||||||||||.|..||.|||.|.||||||||.||||||||..|||||||||||||||||||||||||.||....||||...|||.||||.||||.|.||||..---||||.|||.||||||||||.|||||||||||||||.|||.|||.|.|-.|||.|.|||||||||||||||||.||||||||.|||||||||||.||||||.|||.|.--.|||||||||||||||.||||||||.||.||||.|||||.|||||||||||.||||.||..|||||||||.|||||.||||||||||.|.|||||.|||...||||||||.|||||||||||||||||||||||||||||||||||||||||.||.||||||||---|||||||||||||||||.||||||||||.||||||||||||||||--||..|||||||||||||||.|||||||.|||.|||||||..|||||.|||||||||||||||||.|.||||||-||||.|||||||||..||||||||||||||.||||||..||||||.|||..||.|||.||.||||||||||.||||||.||.|||---||..|||||||||||||||||----------------------------------------------------------------------------------------------------------------||||.|||.|.|.||||||..|||||||||||||.||.||||||||||||||.|||||||||.||||||||||||||||||||.||||..|||||||||.|..|||.||||||..|||||||..|||.||||||||..||||||||---||||.|..|||||||.|.||||||.|||||||||||||||..||||||.--||..||.|||.|||||.|.|||||...||||||.||||||.||||||||||||||||||.|||.||||.|||||.|---|-------...|-----\n",
      "MILKSFILGNLVSLCMKIVNSVVVVGLYYGFLTTFSMGPSYLFLLRARVIEEGEEGTEKKVSATTGFITGQFMMFISIYYAPLHLALGRPHTITVLALPYLLFHFFWNNHKHFFDYGSTTRNSMRNLSIQFVFLNNLIFQLFNHFILPSSMLVRLVNIYMFRCNNKMLFVTSSFVGWLIGHILFMKWVGLILVWIQQNNSIRSNVLFRSNKYLVSELKNSMARIFSILLFITCVYSLGRIPSPIFTKKLKETSETEER---EEETDVEIETTSETKGTKQGSAEEDPSSSLFSEEKEDPDKIDETEEIQVNGKEKTKDEFHFHFKETWYKNRPLYETFYLDGNQENSKLEILIEKKKKNLFWFEKPLVTILFDSKRWNRPFRYIKNDPFQNAVRNEMSQYFFYTCRSDGKENFSFTYPASLSTFLEMVQRKMFLF--TTEKLPSVELYNHWNHKNEQKKKNLSNEFLSRVQALDKGYLVLNTLQKRTRLCNDKTKKEYLPKIYDPLLSGSYRRKILFFFSPSTL--NKTSIKNSTEMVWINKIHLILLISNYIEFEPKADRKSFSTEIAYCLNLINEFAGKSISSFHFKGFPLFPDYKEEKMNLELLQILKFVFDPVLANSKNKTIRKNSTGINGGIKEIKKKVPRWSYKVIEEFEQQETENEENMSENHEIRSRKAKRVVIFTDNQQNTATTKNTKDTTNFDQIDEVTLIRYSQHSDFRRDIIKGSMRAQRRKIAIFELFQANAHSPLFLDRLEKSLFFSFNISELIKTMFINWMCKNTEFRISDSTYTDIEKKTPENKKKDEDKREDDKREERTRIEIAEAWDSILFSQVIRSLILVTQSILRKYIILPSLIIVKNVIRILFFQIPEWSEDLKDWRGEMHVKCTYNGVQLSEKEFPKNWLTDGIQIKILFPFRLKPWHRSKLKSPNEDPLKKKAQKNDFCFLTVWGMEAELPFCSPRKRLSLFEPIFKKLEKKIIKMKKKVFRVITILEERIHFFLNFSKEKKNWVIKNIFFRKEL---INKFSKSKGNRILLSGVKEIYELNATKKEKDSLINNNNRMSQKLSPSMRRSMIWANFSLTEKKMKDLSARRKMILNQIEKIAKEKEKKIITSETNISPNKKSYNTKI--FKSSKNISQILKRRNARLAHKFHFFLKIVIERISIDIFLHIINIPRINAQLCLESTNKILNKYIYNNEENHKKIDKINQNTIHFISILNLKKSLNHSNIAIMNKNSQILCGISSLSQAYVFYKLSQIKIINLYKLRSILQYHGLFLKNEIKDFFRAQGLFNFELKEKKEKNVRNSVMNQWKNWLKSHYQYKYDLSPIRWSRLIPQKWRNRIN--QHSIVQNKKLNKWNLYEKDPLIHYEKKNDFETDSLPNQKKHFKKHYRYNLLSYKSINYENKKDSDISGSPLQINNNKQRISYNYNTNKRKFFDMLEGIPINNYLVEDDIIDREKSPDRKFFDWNMLNLCLRKKVGIESWIDTGSEDKKNTKTNKYKIIE---KKNLFFITIHQDQEVNSSNQKRGGVDWMGMNEEIQKGLISNFELWFFRKFVILYNTYKITPWAIPIQSLLLNCNRNENESKNKQINEKKNGDLFSSISSNETKIQIIELENRNHEEKESEDQVDFGSVFANQEKDIEEDYMGLGMTKHRHKKQKKSHTEVELDFFLKTYLCFQLRWNGSLNQKIIKNIKVYCLLLRLTNPREIIISSIQRQEISLNVLMIQNYLTLTELMKKGILIIEPVRLSVKNDEQFLLYQTVGILLVHKNKQQINKKYRDKIDSISIDSIERHQSLIGNRDKNNYYLLVPENILSPKRRRELRILISFTFKNKNNIP--INIEICNGNNIKNCGPFFQKSKDFYRDKNKLIKLKFFLWPNFRLEDLACMNRYWFDTNNASRFSMVRIYIYPRLSFW*GC-------IFHI-----\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"ycf1\"][\"index\"]\n",
    "name = sacha.features[idx].qualifiers[\"gene\"][0]\n",
    "\n",
    "query = sacha.features[idx].extract(sacha).seq.translate(table=11)\n",
    "\n",
    "for ft in ricinus.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "        target = ft.extract(ricinus).seq.translate(table=11)\n",
    "\n",
    "alns = aligner.align(target,query)\n",
    "print(f\"Alignment and score {alns.score} - target ({len(target)}) - query ({len(query)})\")\n",
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ycf1 presented a case similar to rpl22: Both the reading frame and the start codon were the correct and already had an interior stop codon. Hence, the same procedure was used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = anomalus[\"ycf1\"][\"index\"]\n",
    "alt_loc[\"ycf1\"] = {0:early_stop(sacha.features[idx],sacha, 11)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### __ndhG__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally ndhG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment and score 739.0\n",
      "MDLPGPIHDFLLVFLGSGLLVGGLGVVLLPNPIFSAFSLGFVLVCISLLYILSNSHFVAAAQLLIYVGAINVLIIFAVMFMNDSEYSTDFNLWTIGNGITSLVCTTILFLLMSTILDTSWYGVIWTTKLNQILEQDLISNSQQIGIHLSTDFFLPFELISIILLVALIGAISVARQ-*\n",
      "||||..||||||.|||..|..||||||||.|||.||||||.|||||||.|||||||||||||||||||||||||||||||||.|||...|||||.|||.||||||.|...|...||||||||.|||||.|||||||||||.||||||||||||||||.|||||||||||||.||||-|\n",
      "MDLPALIHDFLLFFLGLVLILGGLGVVLLTNPIYSAFSLGLVLVCISLFYILSNSHFVAAAQLLIYVGAINVLIIFAVMFMNGSEYYKEFNLWTVGNGVTSLVCTSIFVSLITIILDTSWYGIIWTTKTNQILEQDLISNGQQIGIHLSTDFFLPFEFISIILLVALIGAIAVARQ**\n",
      "\n"
     ]
    }
   ],
   "source": [
    "idx = anomalus[\"ndhG\"][\"index\"]\n",
    "name = sacha.features[idx].qualifiers[\"gene\"][0]\n",
    "query = sacha.features[idx].extract(sacha).seq.translate(table=11)\n",
    "\n",
    "for ft in arabidopsis.features:\n",
    "    if ft.type == \"CDS\" and ft.qualifiers[\"gene\"][0] == name:\n",
    "        target = ft.extract(arabidopsis).seq.translate(table=11)\n",
    "\n",
    "alns = aligner.align(target,query)\n",
    "print(f\"Alignment and score {alns.score}\")\n",
    "print(alns[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ndhG had a duplicated stop codon as their only fault."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "alt_loc[\"ndhG\"] = {0:early_stop(sacha.features[291], sacha, 11)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## __Wrap up__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To summarize:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rpoC1': {'start': True, 'cod': 1, 'end': 0, 'index': 62},\n",
       " 'atpF': {'start': True, 'cod': 2, 'end': 13, 'index': 75},\n",
       " 'ndhK': {'start': True, 'cod': 2, 'end': 0, 'index': 99},\n",
       " 'petB': {'start': False, 'cod': 1, 'end': 13, 'index': 168},\n",
       " 'petD': {'start': False, 'cod': 2, 'end': 18, 'index': 170},\n",
       " 'infA': {'start': False, 'cod': 0, 'end': 2, 'index': 178},\n",
       " 'rpl16': {'start': False, 'cod': 1, 'end': 10, 'index': 184},\n",
       " 'rpl22': {'start': True, 'cod': 1, 'end': 5, 'index': 188},\n",
       " 'ycf1': {'start': True, 'cod': 1, 'end': False, 'index': 243},\n",
       " 'ndhI': {'start': True, 'cod': 0, 'end': 0, 'index': 254},\n",
       " 'ndhG': {'start': True, 'cod': 0, 'end': 2, 'index': 256},\n",
       " 'ndhD': {'start': False, 'cod': 0, 'end': True, 'index': 262},\n",
       " 'ccsA': {'start': True, 'cod': 2, 'end': 0, 'index': 264},\n",
       " 'ndhF': {'start': True, 'cod': 1, 'end': 0, 'index': 270},\n",
       " 'ycf1-fragment': {'start': True, 'cod': 2, 'end': 0, 'index': 272}}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anomalus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'petB': {37: None,\n",
       "  46: None,\n",
       "  -41: FeatureLocation(ExactPosition(81013), ExactPosition(81700), strand=1)},\n",
       " 'petD': {17: None,\n",
       "  44: None,\n",
       "  56: None,\n",
       "  -1: None,\n",
       "  -7: None,\n",
       "  -25: FeatureLocation(ExactPosition(82687), ExactPosition(83188), strand=1),\n",
       "  -49: FeatureLocation(ExactPosition(82663), ExactPosition(83188), strand=1)},\n",
       " 'rpl16': {43: None,\n",
       "  52: None,\n",
       "  -8: FeatureLocation(ExactPosition(86871), ExactPosition(87282), strand=-1),\n",
       "  -38: None},\n",
       " 'rpoC1': {0: CompoundLocation([FeatureLocation(ExactPosition(31820), ExactPosition(32252), strand=1), FeatureLocation(ExactPosition(33058), ExactPosition(34675), strand=1)], 'join')},\n",
       " 'atpF': {0: CompoundLocation([FeatureLocation(ExactPosition(43236), ExactPosition(43380), strand=1), FeatureLocation(ExactPosition(44098), ExactPosition(44509), strand=1)], 'join')},\n",
       " 'ndhK': {0: FeatureLocation(ExactPosition(53113), ExactPosition(53791), strand=-1)},\n",
       " 'rpl22': {0: FeatureLocation(ExactPosition(89420), ExactPosition(89858), strand=-1)},\n",
       " 'ycf1': {0: FeatureLocation(ExactPosition(116127), ExactPosition(121821), strand=1)},\n",
       " 'ndhI': {0: FeatureLocation(ExactPosition(126132), ExactPosition(126648), strand=1)},\n",
       " 'ndhG': {0: FeatureLocation(ExactPosition(144525), ExactPosition(144573), strand=-1)},\n",
       " 'ccsA': {0: FeatureLocation(ExactPosition(130536), ExactPosition(131502), strand=-1)},\n",
       " 'ndhF': {0: FeatureLocation(ExactPosition(133620), ExactPosition(135876), strand=1)},\n",
       " 'ycf1-fragment': {0: FeatureLocation(ExactPosition(135840), ExactPosition(137817), strand=-1)}}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_loc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were 15 anomalus cds from which we curated and have new locations for 13 of them. The remaining 2 were ycf1-fragment and infA; the former being a natural fragmentation and the later a suspect for sporious match."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, to make a final genbank file, the cds had to be replaced and infA should be removed. It was also noted that the gene pbf1 is named as psbN and other chloroplast; so it is sensible to rename it to preserve the convention. Additionally, in reference files the cds usually include the product of the annotation. Furthermore, as a result of a heteroplasmy search on the chlroplast, it was found that the only ambiguous nucleotide in the sequence (W at 17517), has enough depth to be disambiguated (it is an A)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But first, the dictionary with the modified locations had to be organized:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "del alt_loc[\"ycf1-fragment\"]\n",
    "for key in alt_loc.keys():\n",
    "    alt_loc[key] = {key2:value for key2, value in alt_loc[key].items() if value != None}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'petB': {-41: FeatureLocation(ExactPosition(81013), ExactPosition(81700), strand=1)},\n",
       " 'petD': {-25: FeatureLocation(ExactPosition(82687), ExactPosition(83188), strand=1),\n",
       "  -49: FeatureLocation(ExactPosition(82663), ExactPosition(83188), strand=1)},\n",
       " 'rpl16': {-8: FeatureLocation(ExactPosition(86871), ExactPosition(87282), strand=-1)},\n",
       " 'rpoC1': {0: CompoundLocation([FeatureLocation(ExactPosition(31820), ExactPosition(32252), strand=1), FeatureLocation(ExactPosition(33058), ExactPosition(34675), strand=1)], 'join')},\n",
       " 'atpF': {0: CompoundLocation([FeatureLocation(ExactPosition(43236), ExactPosition(43380), strand=1), FeatureLocation(ExactPosition(44098), ExactPosition(44509), strand=1)], 'join')},\n",
       " 'ndhK': {0: FeatureLocation(ExactPosition(53113), ExactPosition(53791), strand=-1)},\n",
       " 'rpl22': {0: FeatureLocation(ExactPosition(89420), ExactPosition(89858), strand=-1)},\n",
       " 'ycf1': {0: FeatureLocation(ExactPosition(116127), ExactPosition(121821), strand=1)},\n",
       " 'ndhI': {0: FeatureLocation(ExactPosition(126132), ExactPosition(126648), strand=1)},\n",
       " 'ndhG': {0: FeatureLocation(ExactPosition(144525), ExactPosition(144573), strand=-1)},\n",
       " 'ccsA': {0: FeatureLocation(ExactPosition(130536), ExactPosition(131502), strand=-1)},\n",
       " 'ndhF': {0: FeatureLocation(ExactPosition(133620), ExactPosition(135876), strand=1)}}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt_loc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_sacha_fts = sacha.features.copy()\n",
    "\n",
    "for name, dict_loc in alt_loc.items():\n",
    "    idx = anomalus[name][\"index\"] - 1\n",
    "    for offset, loc in dict_loc.items():\n",
    "        while new_sacha_fts[idx].qualifiers[\"gene\"][0] == name:\n",
    "            if new_sacha_fts[idx].type in {\"gene\", \"CDS\"}:\n",
    "                new_sacha_fts[idx].location = loc\n",
    "            elif new_sacha_fts[idx].type == \"exon\":\n",
    "                num = int(new_sacha_fts[idx].qualifiers[\"number\"][0])\n",
    "                new_sacha_fts[idx].location = loc.parts[num - 1]\n",
    "            elif new_sacha_fts[idx].type == \"intron\":\n",
    "                num = int(new_sacha_fts[idx].qualifiers[\"number\"][0])\n",
    "                start = loc.parts[num -1].end\n",
    "                end = loc.parts[num].start\n",
    "                new_sacha_fts[idx].location = FeatureLocation(start, end, loc.strand)\n",
    "            if \"gene\" not in new_sacha_fts[idx + 1].qualifiers.keys():\n",
    "                idx += 2 ## This was cheating, it works just for this case. Not aimed for general use\n",
    "            idx += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To filter the infA and rename \"pbf1\" to \"psbN\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_sacha_fts_filtered = []\n",
    "for ft in new_sacha_fts:\n",
    "    if \"gene\" in ft.qualifiers.keys():\n",
    "        if ft.qualifiers[\"gene\"][0] != \"infA\":\n",
    "            new_sacha_fts_filtered.append(ft)\n",
    "            if ft.qualifiers[\"gene\"][0] == \"pbf1\":\n",
    "                ft.qualifiers[\"gene\"][0] = \"psbN\"\n",
    "    else:\n",
    "        new_sacha_fts_filtered.append(ft)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to add the translation products to the cds and other qualifiers to be NCBI compliant (add product, translation table) :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "270 ycf1-fragment\n"
     ]
    }
   ],
   "source": [
    "definite_features = []\n",
    "for i,ft in enumerate(new_sacha_fts_filtered):\n",
    "    if ft.type == \"CDS\":\n",
    "        name = ft.qualifiers[\"gene\"][0]\n",
    "        if name == \"ndhD\": ## Remember ndhD has a particular starting site\n",
    "            protein = \"M\" + ft.extract(sacha).seq.translate(table=11)[1:-1]\n",
    "            ft.qualifiers[\"transl_except\"] = [\"(pos:1..3,aa:Met)\"]\n",
    "        #elif name == \"ycf1-fragment\":\n",
    "        #    continue\n",
    "        else:\n",
    "            try:\n",
    "                protein = ft.extract(sacha).seq.translate(table=11, cds=True)\n",
    "            except:\n",
    "                print(i, name)\n",
    "                ft.qualifiers[\"exception\"] = \"Opposite IR fragment of protein ycf1\"\n",
    "                protein = ft.extract(sacha).seq.translate(table=11, to_stop=True)\n",
    "                \n",
    "        ## Look for the description in ricinus chloroplast\n",
    "        for feat in ricinus.features:\n",
    "            if feat.type == \"CDS\" and feat.qualifiers[\"gene\"][0] == name:\n",
    "                prod = feat.qualifiers[\"gene\"][0]\n",
    "        \n",
    "        ft.qualifiers[\"product\"] = [prod]\n",
    "        ft.qualifiers[\"translation\"] = [str(protein)]\n",
    "        ft.qualifiers[\"transl_table\"] = [\"11\"]\n",
    "        definite_features.append(ft)\n",
    "    else:\n",
    "        definite_features.append(ft)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally the genbank file is created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb = SeqIO.read(\"chloroplast_annotated_raw.gb\", \"gb\")\n",
    "gb.features = definite_features\n",
    "gb.seq = Seq(str(gb.seq).replace(\"W\",\"A\"), gb.seq.alphabet)\n",
    "SeqIO.write(gb, \"chloroplast_annotated_curated.gb\", \"gb\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Sci",
   "language": "python",
   "name": "sci"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
